db/system_keyspace.cc

```diff
@@ -264,6 +264,9 @@ schema_ptr system_keyspace::topology() {
             .with_column("new_cdc_generation_data_uuid", timeuuid_type, column_kind::static_column)
             .with_column("new_keyspace_rf_change_ks_name", utf8_type, column_kind::static_column)
             .with_column("new_keyspace_rf_change_data", map_type_impl::get_instance(utf8_type, utf8_type, false), column_kind::static_column)
+            .with_column("truncate_table_ks_name", utf8_type, column_kind::static_column)
+            .with_column("truncate_table_cf_name", utf8_type, column_kind::static_column)
+            .with_column("truncate_table_timeout_ms", long_type, column_kind::static_column)
```

Comment:
wouldn't that limit truncating multiple tables in parallel?

Reply:
Yes, it would. We can not run truncate in parallel with another truncate, or any other global topology operation. We could alleviate this by waiting for the operation to complete (we currently fail truncate immediately if there is an ongoing global topology request).
We could also revert to making truncate a tablet transition instead of a global topology request, but that would create problems because truncate is implemented as a table level operation. With the current truncate implementation, we can only issue RPCs to truncate all the tablets of a table on a host; we can not truncate a single tablet. I think this makes implementing truncate as a tablet transition an unnatural fit. But we could make that compromise to achieve parallel truncates.

Reply:
okay, we can live with that for now

Reply:
why not truncate_table_id (with uuid)?

Reply:
isn't it a problem for the long run that we decided to pick system topology for truncate? I can understand keyspace rf change, since it has to do with replicas, but truncate is different (it's an operation on the table, has nothing to do with topology layout). this opens a precedent for putting more stuff here.

Reply:
> why not truncate_table_id (with uuid)?

Agree that it would be safer. Scenario: 1) truncate request queued 2) times out on client side 3) user decides to drop& create the table, and populate it with new data 4) truncate is executed on the new table

Reply:
> why not truncate_table_id (with uuid)?

I agree. I will change that in the next version.

Reply:
> isn't it a problem for the long run that we decided to pick system topology for truncate? I can understand keyspace rf change, since it has to do with replicas, but truncate is different (it's an operation on the table, has nothing to do with topology layout). this opens a precedent for putting more stuff here.

I can see that it is somewhat of a problem. But the place I chose stems from the fact that we decided to make truncate a topology operation (which it really isn't), and we need to pass the truncate parameters to the topology coordinator somehow.
Can you, please, suggest an alternative place for these?

Reply:
@gleb-cloudius do we have a better alternative than putting stuff here?

Reply:
That was never a good alternative. But this is how @avikivity wants it to be as discussed here: https://github.com/scylladb/scylla-enterprise/discussions/4796

I shudder to think what have to be done now to move to multi element global request queue with all this static columns around. Multi element global request queue shouldn't have to be so low priority.

Reply:
Yes it's very ugly.

What's needed is a variant of structs, and I admit there isn't a good tabular representation for it.

Reply:
I am not an expert in this area, so take it with a grain of salt, but wouldn't a better alternative be system.topology_requests? Then each request has a map column with a set of params that are only understood for a given request type. It may be more extensible this way. Options can be deprecated, and new ones can be added.

Reply:
It is more extensible (and people suggested json which is even more extensible). The problem is it creates a new mini-schema which has to be separately documented, and mini-parsers to convert the map value type to the real data type.

Reply:
We already have a representation of an arbitrary object, it's called mutation. Indeed, that's what we store in system.paxos. The schema of the mutation is defined elsewhere.

Reply:
Generally I'd move to the global request queue first, and then rebase this patch on top of it.

Reply:
Only a single parameter has remained and it has been moved to system.topology_requests in v7